### PR TITLE
Ignore cached_property in method-hidden check (#8753)

### LINKED ISSUE #8753: False Positive: method-hidden with @functools.cached_property
### Bug description

When decorating a method with `@functools.cached_property` pylint produces the error `method-hidden` whereas decorating it with `@property` does not.

example.py:
```python
# pylint: disable=too-few-public-methods
"""A bug report example."""
import functools


class Base:
    """Generic base class."""

    def __init__(self, example):
        self.example = example


class Extended(Base):
    """Class showing the method-hidden error."""

    @functools.cached_property
    def example(self):
        """
        This is a property and should be cached, but should not product method-hidden.
        """
        return "expensive function"
```

### Configuration

_No response_

### Command used

```shell
pylint example.py
```

### Pylint output

```shell
************* Module example
example.py:17:4: E0202: An attribute defined in example line 10 hides this method (method-hidden)

------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: 0.00/10, +2.86)
```

### Expected behavior

The `method-hidden` error should not be emitted.

### Pylint version

```shell
pylint 2.15.8
astroid 2.12.13
Python 3.10.6 (main, Aug 31 2022, 16:28:02) [Clang 13.1.6 (clang-1316.0.21.2.5)]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by kyoto7250**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description
Closes #8753

### CODE REVIEW

File: tests/functional/m/method_hidden.py
Comment (Pierre-Sassoulas) [inline tests/functional/m/method_hidden.py:5]:
```suggestion
import functools as ft
import something_else as functools
```

File: tests/functional/m/method_hidden.py
Comment (Pierre-Sassoulas) [inline tests/functional/m/method_hidden.py:118]:
```suggestion
    @ft.cached_property
```

File: tests/functional/m/method_hidden.py
Comment (Pierre-Sassoulas) [inline tests/functional/m/method_hidden.py:121]:
```suggestion

    @functools.cached_property
        def _protected_two(self):
            pass
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks pretty nice already ! I suggested a test case to make sure we use the qname / inference, not sure if this already work or how necessary this is.

File: tests/functional/m/method_hidden.py
Comment (kyoto7250) [inline tests/functional/m/method_hidden.py:126]: There is no actual definition `cached_property`, so `safe_infer` returns `UnInferable`, and this line should have been pointed out by the pylint.

By adding the actual definition, I think this line will be detected, any good ideas?

https://github.com/pylint-dev/pylint/blob/1ef3b0be3c018ab4f2928932ab1f46ee244b79bb/pylint/checkers/classes/class_checker.py#L1261-L1262

Comment (Pierre-Sassoulas) [reply tests/functional/m/method_hidden.py:126]: It's the expected behavior 👍 We should not raise when something is uninferable (unless the real functools.cached_property is always uninferable).

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:52]:
```suggestion
ALLOWED_PROPERTIES = {"property", "cached_property"}
```

Or using ``qname``: 
```suggestion
ALLOWED_PROPERTIES = {"bultins.property", "functools.cached_property"}
```

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1256]:
```suggestion
                    if decorator.name in ALLOWED_PROPERTIES:
```

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1464]: What if we have some import that is ``from functools import cached_property as cached`` ? But let's hear from someone else first, I'm not sure handling alias perfectly is that important, aliasing doesn't happen often and it's bad practice imo.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for taking the remark into account. Let's wait for another maintainer opinion regarding aliasing before investing too much time in it :) If it's important then we should use ``qname`` (qualified name, i.e. ``builtins.property``) instead.

File: pylint/checkers/classes/class_checker.py
Comment (jacobtylerwalls) [inline pylint/checkers/classes/class_checker.py:1467]: This line is easily coverable by testing with a decorator that's a Call, like `@cached_property()`

File: pylint/checkers/classes/class_checker.py
Comment (jacobtylerwalls) [reply pylint/checkers/classes/class_checker.py:1464]: Yeah, I let's leave that out of scope. We didn't even do that for `TYPE_CHECKING` even though we did some elaborate work for aliasing `typing`.

Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:1464]: Not sure about it but we could revert 8d62ba0a0b55f24d010082b8e8f484ab7d816f37 then ?

Comment (jacobtylerwalls) [reply pylint/checkers/classes/class_checker.py:1464]: Nah, I think we need it because it makes the aliasing functools as ft test case work.

**Comment by jacobtylerwalls** [comment]
The docs on adding a changelog are [here](https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/contribute.html) (you can install the dev requirements to get `towncrier`).

**Comment by kyoto7250** [comment]
Thank you for your reviews.

I commited the changes and create a news fragment.

File: pylint/checkers/classes/class_checker.py
Comment (jacobtylerwalls) [reply pylint/checkers/classes/class_checker.py:1467]: @kyoto7250 I think this line is still missing coverage. You can add a test case with a decorator that's a call. Probably better not to use `cached_property()` as the fake call, since that could be confusing.

Comment (kyoto7250) [reply pylint/checkers/classes/class_checker.py:1467]: I tried https://github.com/pylint-dev/pylint/pull/8758/commits/a3827965c6e1461f5c0859a4da55cafb6f26666e, but This doesn't seem to be a valid in python3.8.

The test here is a little difficult. `L1261` confirms whether it is an `nodes.Attribute`, so I think that this conditional branch will not be entered in if we only use `nodes.Call`.

Comment (jacobtylerwalls) [reply pylint/checkers/classes/class_checker.py:1467]: That's actually the correct test, and if you want to verify the line is executed, just insert a bullet like `1 / 0` before the return and watch it fail.

> This doesn't seem to be a valid in python3.8.

That's okay! Seems to be a legitimate syntax difference only enabled on 3.9+ and up. The test case is still a good one, you'll just need to move it to a new file suffixed with _py39.py and with a corresponding file ending in `.rc` with the min-version set to py 3.9. `method-cache-max-size-none` has tests set up exactly so.

Comment (kyoto7250) [reply pylint/checkers/classes/class_checker.py:1467]: Thank you for your comment.
I commited https://github.com/pylint-dev/pylint/pull/8758/commits/70615c29f520a45ecd70ae2b6ffa080d73bb5792

File: tests/functional/m/method_hidden_py39.py
Comment (jacobtylerwalls) [inline tests/functional/m/method_hidden_py39.py:22]: You can actually delete most of this file. It only needs to contain the one test that doesn't run on Python 3.8. All the other tests in the other file will still run on all versions.

(This also has the benefit of clarifying which tests have to do purely with your PR.)

File: tests/functional/m/method_hidden.py
Comment (jacobtylerwalls) [inline tests/functional/m/method_hidden.py:3]: Cosmetic suggestion: I'd suggest disabling this on line 6 only, not here for the entire file.

**Comment by jacobtylerwalls** [comment]
Sorry about the unhelpful primer message. I'll open an unrelated issue for it.

File: pylint/checkers/classes/class_checker.py
Comment (jacobtylerwalls) [reply pylint/checkers/classes/class_checker.py:1472]: However, those would make pretty low-value tests, so I'm approving now with an eye toward merging later tonight. Thank you!

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
Thanks for being so responsive @kyoto7250, great first contribution! 📈

**Pull request merged**
